File: PolymarketLab.Framework/Response/Envelope.cs
namespace PolymarketLab.Framework.Response;

public record ResponseError(string? ErrorCode, string? ErrorMessage, string? InvalidField);

/// <summary>
///     Шаблон ответа
/// </summary>
public record Envelope
{
    public Envelope(object? result, IEnumerable<ResponseError> errors)
    {
        Result = result;
        ListErrors = errors.ToList();
        CreatedOtc = DateTime.UtcNow;
    }

    /// <summary>
    ///     Результат если есть
    /// </summary>
    public object? Result { get; }

    /// <summary>
    ///     Если есть ошибка,то статус код ошибки
    /// </summary>
    public List<ResponseError> ListErrors { get; }

    /// <summary>
    ///     Дата запроса
    /// </summary>
    public DateTime CreatedOtc { get; }

    public static Envelope Ok(object? result = null)
    {
        return new Envelope(result, []);
    }

    public static Envelope Errors(IEnumerable<ResponseError> errors)
    {
        return new Envelope(null, errors);
    }
}

File: PolymarketLab.Web/src/api/envelope.ts
export interface ResponseError {
  errorCode: string | null;
  errorMessage: string | null;
  invalidField: string | null;
}

export interface Envelope<TResult> {
  result: TResult | null;
  listErrors: ResponseError[];
  createdOtc: string;
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

export function isResponseError(value: unknown): value is ResponseError {
  if (!isObject(value)) {
    return false;
  }

  return (
    ('errorCode' in value &&
      (value.errorCode === null || typeof value.errorCode === 'string')) &&
    ('errorMessage' in value &&
      (value.errorMessage === null || typeof value.errorMessage === 'string')) &&
    ('invalidField' in value &&
      (value.invalidField === null || typeof value.invalidField === 'string'))
  );
}

export function isEnvelope<TResult>(value: unknown): value is Envelope<TResult> {
  if (!isObject(value)) {
    return false;
  }

  return (
    'result' in value &&
    'listErrors' in value &&
    Array.isArray(value.listErrors) &&
    value.listErrors.every(isResponseError) &&
    'createdOtc' in value &&
    typeof value.createdOtc === 'string'
  );
}

export function getResponseErrorMessage(
  errors: ResponseError[],
  fallbackMessage: string,
): string {
  for (const error of errors) {
    const message = error.errorMessage?.trim();
    if (message) {
      return message;
    }

    const code = error.errorCode?.trim();
    if (code) {
      return code;
    }
  }

  return fallbackMessage;
}
